# Machine Learning With Spark ML
In this lab assignment, you will complete a project by going through the following steps:
1. Get the data.
2. Discover the data to gain insights.
3. Prepare the data for Machine Learning algorithms.
4. Select a model and train it.
5. Fine-tune your model.
6. Present your solution.

As a dataset, we use the California Housing Prices dataset from the StatLib repository. This dataset was based on data from the 1990 California census. The dataset has the following columns
1. `longitude`: a measure of how far west a house is (a higher value is farther west)
2. `latitude`: a measure of how far north a house is (a higher value is farther north)
3. `housing_,median_age`: median age of a house within a block (a lower number is a newer building)
4. `total_rooms`: total number of rooms within a block
5. `total_bedrooms`: total number of bedrooms within a block
6. `population`: total number of people residing within a block
7. `households`: total number of households, a group of people residing within a home unit, for a block
8. `median_income`: median income for households within a block of houses
9. `median_house_value`: median house value for households within a block
10. `ocean_proximity`: location of the house w.r.t ocean/sea

---
# 1. Get the data
Let's start the lab by loading the dataset. The can find the dataset at `data/housing.csv`. To infer column types automatically, when you are reading the file, you need to set `inferSchema` to true. Moreover enable the `header` option to read the columns' name from the file.

In [1]:
// TODO: Replace <FILL IN> with appropriate code

val housing = spark.read.format("csv")
              .option("header", "true")
              .option("inferSchema", "true")
              .load("data/housing.csv")

housing = [longitude: double, latitude: double ... 8 more fields]


[longitude: double, latitude: double ... 8 more fields]

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting
Predicting

---
# 2. Discover the data to gain insights
Now it is time to take a look at the data. In this step we are going to take a look at the data a few different ways:
* See the schema and dimension of the dataset
* Look at the data itself
* Statistical summary of the attributes
* Breakdown of the data by the categorical attribute variable
* Find the correlation among different attributes
* Make new attributes by combining existing attributes

## 2.1. Schema and dimension
Print the schema of the dataset

In [2]:
// TODO: Replace <FILL IN> with appropriate code

housing.printSchema()

root
 |-- longitude: double (nullable = true)
 |-- latitude: double (nullable = true)
 |-- housing_median_age: double (nullable = true)
 |-- total_rooms: double (nullable = true)
 |-- total_bedrooms: double (nullable = true)
 |-- population: double (nullable = true)
 |-- households: double (nullable = true)
 |-- median_income: double (nullable = true)
 |-- median_house_value: double (nullable = true)
 |-- ocean_proximity: string (nullable = true)



Print the number of records in the dataset.

In [3]:
// TODO: Replace <FILL IN> with appropriate code

housing.count()

20640

## 2.2. Look at the data
Print the first five records of the dataset.

In [4]:
// TODO: Replace <FILL IN> with appropriate code

housing.head(5)

Array([-122.23,37.88,41.0,880.0,129.0,322.0,126.0,8.3252,452600.0,NEAR BAY], [-122.22,37.86,21.0,7099.0,1106.0,2401.0,1138.0,8.3014,358500.0,NEAR BAY], [-122.24,37.85,52.0,1467.0,190.0,496.0,177.0,7.2574,352100.0,NEAR BAY], [-122.25,37.85,52.0,1274.0,235.0,558.0,219.0,5.6431,341300.0,NEAR BAY], [-122.25,37.85,52.0,1627.0,280.0,565.0,259.0,3.8462,342200.0,NEAR BAY])

Print the number of records with population more than 10000.

In [5]:
// TODO: Replace <FILL IN> with appropriate code

housing.where("population > 10000").count()

23

## 2.3. Statistical summary
Print a summary of the table statistics for the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. You can use the `describe` command.

In [6]:
// TODO: Replace <FILL IN> with appropriate code

housing.describe("housing_median_age", "total_rooms", "median_house_value", "population").show()

+-------+------------------+------------------+------------------+------------------+
|summary|housing_median_age|       total_rooms|median_house_value|        population|
+-------+------------------+------------------+------------------+------------------+
|  count|             20640|             20640|             20640|             20640|
|   mean|28.639486434108527|2635.7630813953488|206855.81690891474|1425.4767441860465|
| stddev| 12.58555761211163|2181.6152515827944|115395.61587441359|  1132.46212176534|
|    min|               1.0|               2.0|           14999.0|               3.0|
|    max|              52.0|           39320.0|          500001.0|           35682.0|
+-------+------------------+------------------+------------------+------------------+



Print the maximum age (`housing_median_age`), the minimum number of rooms (`total_rooms`), and the average of house values (`median_house_value`).

In [7]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.sql.functions._

housing.select(max("housing_median_age"), min("total_rooms"), avg("median_house_value")).show()

+-----------------------+----------------+-----------------------+
|max(housing_median_age)|min(total_rooms)|avg(median_house_value)|
+-----------------------+----------------+-----------------------+
|                   52.0|             2.0|     206855.81690891474|
+-----------------------+----------------+-----------------------+



## 2.4. Breakdown the data by categorical data
Print the number of houses in different areas (`ocean_proximity`), and sort them in descending order.

In [8]:
// TODO: Replace <FILL IN> with appropriate code

housing.groupBy("ocean_proximity").count().orderBy(desc("count")).show()

+---------------+-----+
|ocean_proximity|count|
+---------------+-----+
|      <1H OCEAN| 9136|
|         INLAND| 6551|
|     NEAR OCEAN| 2658|
|       NEAR BAY| 2290|
|         ISLAND|    5|
+---------------+-----+



Print the average value of the houses (`median_house_value`) in different areas (`ocean_proximity`), and call the new column `avg_value` when print it.

In [9]:
// TODO: Replace <FILL IN> with appropriate code

housing.groupBy("ocean_proximity").agg(avg("median_house_value").alias("avg_value")).show()

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|     NEAR OCEAN|249433.97742663656|
|       NEAR BAY|259212.31179039303|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



Rewrite the above question in SQL.

In [10]:
// TODO: Replace <FILL IN> with appropriate code

housing.createOrReplaceTempView("df")
spark.sql("SELECT ocean_proximity, avg(median_house_value) AS avg_value FROM df GROUP BY ocean_proximity ORDER BY avg_value DESC").show()

+---------------+------------------+
|ocean_proximity|         avg_value|
+---------------+------------------+
|         ISLAND|          380440.0|
|       NEAR BAY|259212.31179039303|
|     NEAR OCEAN|249433.97742663656|
|      <1H OCEAN|240084.28546409807|
|         INLAND|124805.39200122119|
+---------------+------------------+



## 2.5. Correlation among attributes
Print the correlation among the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. To do so, first you need to put these attributes into one vector. Then, compute the standard correlation coefficient (Pearson) between every pair of attributes in this new vector. To make a vector of these attributes, you can use the `VectorAssembler` Transformer.

In [11]:
// TODO: Replace <FILL IN> with appropriate code


import org.apache.spark.ml.feature.VectorAssembler

// get input multiple columns, set as output the name of the column output that will represent all the features of each row 
// (spark row formatting [a,b,c,d,...])
val va = new VectorAssembler()
             .setInputCols(Array("housing_median_age", "total_rooms", "median_house_value", "population"))
             .setOutputCol("features")

val housingAttrs = va.transform(housing)

housingAttrs.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+--------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|            features|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+--------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|[41.0,880.0,45260...|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|          358500.0|       NEAR BAY|[21.0,7099.0,3585...|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|          352100.0|       NEAR BAY|[52.0,1467.0,3521...|
|  -122.25|   37.85|              52.0|     12

va = vecAssembler_0be63ccd1235
housingAttrs = [longitude: double, latitude: double ... 9 more fields]


[longitude: double, latitude: double ... 9 more fields]

In [12]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.linalg.Matrix
import org.apache.spark.ml.stat.Correlation
import org.apache.spark.sql.Row

val Row(coeff: Matrix) = Correlation.corr(housingAttrs, "features").head

println(s"The standard correlation coefficient:\n ${coeff}")

The standard correlation coefficient:
 1.0                   -0.36126220122231784  0.10562341249318154   -0.2962442397735293   
-0.36126220122231784  1.0                   0.13415311380654338   0.8571259728659772    
0.10562341249318154   0.13415311380654338   1.0                   -0.02464967888891235  
-0.2962442397735293   0.8571259728659772    -0.02464967888891235  1.0                   


coeff = 


1.0                   -0.36126220122231784  0.10562341249318154   -0.2962442397735293
-0.36126220122231784  1.0                   0.13415311380654338   0.8571259728659772
0.10562341249318154   0.13415311380654338   1.0                   -0.02464967888891235
-0.2962442397735293   0.8571259728659772    -0.02464967888891235  1.0


## 2.6. Combine and make new attributes
Now, let's try out various attribute combinations. In the given dataset, the total number of rooms in a block is not very useful, if we don't know how many households there are. What we really want is the number of rooms per household. Similarly, the total number of bedrooms by itself is not very useful, and we want to compare it to the number of rooms. And the population per household seems like also an interesting attribute combination to look at. To do so, add the three new columns to the dataset as below. We will call the new dataset the `housingExtra`.
```
rooms_per_household = total_rooms / households
bedrooms_per_room = total_bedrooms / total_rooms
population_per_household = population / households
```

In [13]:
// TODO: Replace <FILL IN> with appropriate code

val housingCol1 = housing.withColumn("rooms_per_household", expr("total_rooms / households"))
val housingCol2 = housingCol1.withColumn("bedrooms_per_room", expr("total_bedrooms / total_rooms"))
val housingExtra = housingCol2.withColumn("population_per_household", expr("population / households"))

housingExtra.select("rooms_per_household", "bedrooms_per_room", "population_per_household").show(5)

+-------------------+-------------------+------------------------+
|rooms_per_household|  bedrooms_per_room|population_per_household|
+-------------------+-------------------+------------------------+
|  6.984126984126984|0.14659090909090908|      2.5555555555555554|
|  6.238137082601054|0.15579659106916466|       2.109841827768014|
|  8.288135593220339|0.12951601908657123|      2.8022598870056497|
| 5.8173515981735155|0.18445839874411302|       2.547945205479452|
|  6.281853281853282| 0.1720958819913952|      2.1814671814671813|
+-------------------+-------------------+------------------------+
only showing top 5 rows



housingCol1 = [longitude: double, latitude: double ... 9 more fields]
housingCol2 = [longitude: double, latitude: double ... 10 more fields]
housingExtra = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

---
## 3. Prepare the data for Machine Learning algorithms
Before going through the Machine Learning steps, let's first rename the label column from `median_house_value` to `label`.

In [14]:
// TODO: Replace <FILL IN> with appropriate code

val renamedHousing = housingExtra.withColumnRenamed("median_house_value","label")

renamedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

Now, we want to separate the numerical attributes from the categorical attribute (`ocean_proximity`) and keep their column names in two different lists. Moreover, sice we don't want to apply the same transformations to the predictors (features) and the label, we should remove the label attribute from the list of predictors. 

In [15]:
// label columns
val colLabel = "label"

// categorical columns
val colCat = "ocean_proximity"

// numerical columns
val colNum = renamedHousing.columns.filter(_ != colLabel).filter(_ != colCat)

colLabel = label
colCat = ocean_proximity
colNum = Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)


Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household)

## 3.1. Prepare continuse attributes
### Data cleaning
Most Machine Learning algorithms cannot work with missing features, so we should take care of them. As a first step, let's find the columns with missing values in the numerical attributes. To do so, we can print the number of missing values of each continues attributes, listed in `colNum`.

In [16]:
// TODO: Replace <FILL IN> with appropriate code

for (c <- colNum) {
    var count = renamedHousing.filter(renamedHousing(c).isNull || renamedHousing(c) === "" || renamedHousing(c).isNaN).count()
    println(s"$c column has $count missing values\n")
}

longitude column has 0 missing values

latitude column has 0 missing values

housing_median_age column has 0 missing values

total_rooms column has 0 missing values

total_bedrooms column has 207 missing values

population column has 0 missing values

households column has 0 missing values

median_income column has 0 missing values

rooms_per_household column has 0 missing values

bedrooms_per_room column has 207 missing values

population_per_household column has 0 missing values



As we observerd above, the `total_bedrooms` and `bedrooms_per_room` attributes have some missing values. One way to take care of missing values is to use the `Imputer` Transformer, which completes missing values in a dataset, either using the mean or the median of the columns in which the missing values are located. To use it, you need to create an `Imputer` instance, specifying that you want to replace each attribute's missing values with the "median" of that attribute.

In [17]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.Imputer

val imputer = new Imputer().setStrategy("median")
                           .setInputCols(Array("total_bedrooms", "bedrooms_per_room"))
                           .setOutputCols(Array("total_bedrooms", "bedrooms_per_room")) 
                         // columns to input      -    output to write (in this case I re-write them)
val imputedHousing = imputer.fit(renamedHousing).transform(renamedHousing)

imputedHousing.select("total_bedrooms", "bedrooms_per_room").show(5)

+--------------+-------------------+
|total_bedrooms|  bedrooms_per_room|
+--------------+-------------------+
|         129.0|0.14659090909090908|
|        1106.0|0.15579659106916466|
|         190.0|0.12951601908657123|
|         235.0|0.18445839874411302|
|         280.0| 0.1720958819913952|
+--------------+-------------------+
only showing top 5 rows



imputer = imputer_ff8c0b0266fc
imputedHousing = [longitude: double, latitude: double ... 11 more fields]


[longitude: double, latitude: double ... 11 more fields]

### Scaling
One of the most important transformations you need to apply to your data is feature scaling. With few exceptions, Machine Learning algorithms don't perform well when the input numerical attributes have very different scales. This is the case for the housing data: the total number of rooms ranges from about 6 to 39,320, while the median incomes only range from 0 to 15. Note that scaling the label attribues is generally not required.

One way to get all attributes to have the same scale is to use standardization. In standardization, for each value, first it subtracts the mean value (so standardized values always have a zero mean), and then it divides by the variance so that the resulting distribution has unit variance. To do this, we can use the `StandardScaler` Estimator. To use `StandardScaler`, again we need to convert all the numerical attributes into a big vectore of features using `VectorAssembler`, and then call `StandardScaler` on that vactor.

In [18]:
import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

val va = new VectorAssembler()
             .setInputCols(colNum)
             .setOutputCol("features")

val featuredHousing = va.transform(imputedHousing)

val scaler = new StandardScaler()
                 .setInputCol("features")
                 .setOutputCol("scaledFeatures")
val scaledHousing = scaler.fit(featuredHousing).transform(featuredHousing)

scaledHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|            features|      scaledFeatures|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|[-122.23,37.88,41...|[-61.007269596069...|
|  -122.22|   37.86|              21.0|     7099.0|        1

va = vecAssembler_49012e097446
featuredHousing = [longitude: double, latitude: double ... 12 more fields]
scaler = stdScal_68dc8dc079c6
scaledHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

## 3.2. Prepare categorical attributes
After imputing and scaling the continuse attributes, we should take care of the categorical attributes. Let's first print the number of distict values of the categirical attribute `ocean_proximity`.

In [19]:
// TODO: Replace <FILL IN> with appropriate code

renamedHousing.select(countDistinct("ocean_proximity")).show

+-------------------------------+
|count(DISTINCT ocean_proximity)|
+-------------------------------+
|                              5|
+-------------------------------+



### String indexer
Most Machine Learning algorithms prefer to work with numbers. So let's convert the categorical attribute `ocean_proximity` to numbers. To do so, we can use the `StringIndexer` that encodes a string column of labels to a column of label indices. The indices are in [0, numLabels), ordered by label frequencies, so the most frequent label gets index 0.

In [20]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.StringIndexer

// "Smart" One hot encoder, orders based on label frequencu (0 if most frequent, 1 the second most fre, ... )
val indexer = new StringIndexer()
                  .setInputCol("ocean_proximity")
                  .setOutputCol("ocean_proximity_index")
val idxHousing = indexer.fit(renamedHousing).transform(renamedHousing)

idxHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+---------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|ocean_proximity_index|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+---------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|                  3.0|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|358500.0|       NEAR BAY|  6.238137082

indexer = strIdx_8e51fe19310e
idxHousing = [longitude: double, latitude: double ... 12 more fields]


[longitude: double, latitude: double ... 12 more fields]

Now we can use this numerical data in any Machine Learning algorithm. You can look at the mapping that this encoder has learned using the `labels` method: "<1H OCEAN" is mapped to 0, "INLAND" is mapped to 1, etc.

In [21]:
indexer.fit(renamedHousing).labels

Array(<1H OCEAN, INLAND, NEAR OCEAN, NEAR BAY, ISLAND)

### One-hot encoding
Now, convert the label indices built in the last step into one-hot vectors. To do this, you can take advantage of the `OneHotEncoderEstimator` Estimator.

In [22]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.feature.OneHotEncoderEstimator

val encoder = new OneHotEncoderEstimator()
                  .setInputCols(Array("ocean_proximity_index"))
                  .setOutputCols(Array("ocean_proximity_vec"))
val ohHousing = encoder.fit(idxHousing).transform(idxHousing)

ohHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+---------------------+-------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|ocean_proximity_index|ocean_proximity_vec|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+---------------------+-------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|      2.5555555555555554|                  3.0|      (4,[3],[1.0])|
|  -122.22|   37.86|              21.0|     7099.0|        1

encoder = oneHotEncoder_3350b6adda62
ohHousing = [longitude: double, latitude: double ... 13 more fields]


[longitude: double, latitude: double ... 13 more fields]

---
# 4. Pipeline
As you can see, there are many data transformation steps that need to be executed in the right order. For example, you called the `Imputer`, `VectorAssembler`, and `StandardScaler` from left to right. However, we can use the `Pipeline` class to define a sequence of Transformers/Estimators, and run them in order. A `Pipeline` is an `Estimator`, thus, after a Pipeline's `fit()` method runs, it produces a `PipelineModel`, which is a `Transformer`.

Now, let's create a pipeline called `numPipeline` to call the numerical transformers you built above (`imputer`, `va`, and `scaler`) in the right order from left to right, as well as a pipeline called `catPipeline` to call the categorical transformers (`indexer` and `encoder`). Then, put these two pipelines `numPipeline` and `catPipeline` into one pipeline.

In [23]:
// call the preprocessing steps for running SparkML algorithm at once
import org.apache.spark.ml.{Pipeline, PipelineModel, PipelineStage}

val numPipeline = new Pipeline().setStages(Array(imputer, va, scaler))
val catPipeline = new Pipeline().setStages(Array(indexer, encoder))
val pipeline = new Pipeline().setStages(Array(numPipeline, catPipeline))
val newHousing = pipeline.fit(renamedHousing).transform(renamedHousing)

newHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+---------------------+-------------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|   label|ocean_proximity|rooms_per_household|  bedrooms_per_room|population_per_household|            features|      scaledFeatures|ocean_proximity_index|ocean_proximity_vec|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+---------------------+-------------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|452600.0|       NEAR BAY|  6.984126984126984|0.14659090909090908|  

numPipeline = pipeline_ac9c35bf15bc
catPipeline = pipeline_4fcaaa5ffb03
pipeline = pipeline_132d30f564b9
newHousing = [longitude: double, latitude: double ... 15 more fields]


[longitude: double, latitude: double ... 15 more fields]

Now, use `VectorAssembler` to put all attributes of the final dataset `newHousing` into a big vector, and call the new column `features`.

In [24]:
// numerical columns
val finalHousing = newHousing.drop(colCat, "ocean_proximity_index", "features", "scaledFeatures")
finalHousing.printSchema()
val newColNum = finalHousing.columns.filter(_ != colLabel).filter(_ != colCat)

val va2 = new VectorAssembler()
              .setInputCols(newColNum)
              .setOutputCol("to_be_scaled_features")

val featuredHousing2 = va2.transform(finalHousing)

val scaler2 = new StandardScaler()
                 .setInputCol("to_be_scaled_features")
                 .setOutputCol("features")

val dataset = scaler2.fit(featuredHousing2).transform(featuredHousing2).select("features", "label")

dataset.show(5)

root
 |-- longitude: double (nullable = true)
 |-- latitude: double (nullable = true)
 |-- housing_median_age: double (nullable = true)
 |-- total_rooms: double (nullable = true)
 |-- total_bedrooms: double (nullable = true)
 |-- population: double (nullable = true)
 |-- households: double (nullable = true)
 |-- median_income: double (nullable = true)
 |-- label: double (nullable = true)
 |-- rooms_per_household: double (nullable = true)
 |-- bedrooms_per_room: double (nullable = true)
 |-- population_per_household: double (nullable = true)
 |-- ocean_proximity_vec: vector (nullable = true)

+--------------------+--------+
|            features|   label|
+--------------------+--------+
|[-61.007269596069...|452600.0|
|[-61.002278409814...|358500.0|
|[-61.012260782324...|352100.0|
|[-61.017251968579...|341300.0|
|[-61.017251968579...|342200.0|
+--------------------+--------+
only showing top 5 rows



finalHousing = [longitude: double, latitude: double ... 11 more fields]
newColNum = Array(longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, rooms_per_household, bedrooms_per_room, population_per_household, ocean_proximity_vec)
va2 = vecAssembler_5283e27b338e
featuredHousing2 = [longitude: double, latitude: double ... 12 more fields]
scaler2 = stdScal_970820400c32
dataset = [features: vector, label: double]


[features: vector, label: double]

---
# 5. Make a model
Here we going to make four different regression models:
* Linear regression model
* Decission tree regression
* Random forest regression
* Gradient-booster forest regression

But, before giving the data to train a Machine Learning model, let's first split the data into training dataset (`trainSet`) with 80% of the whole data, and test dataset (`testSet`) with 20% of it.

In [25]:
// TODO: Replace <FILL IN> with appropriate code

val Array(trainSet, testSet) = dataset.randomSplit(Array(0.8,0.2))

trainSet = [features: vector, label: double]
testSet = [features: vector, label: double]


[features: vector, label: double]

## 5.1. Linear regression model
Now, train a Linear Regression model using the `LinearRegression` class. Then, print the coefficients and intercept of the model, as well as the summary of the model over the training set by calling the `summary` method.

In [26]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.LinearRegression

// train the model
val lr = new LinearRegression()
    .setMaxIter(100)
    .setRegParam(0)
    .setElasticNetParam(1)
val lrModel = lr.fit(trainSet)
val trainingSummary = lrModel.summary

println(s"Coefficients: ${lrModel.coefficients}, Intercept: ${lrModel.intercept}")
println(s"RMSE: ${trainingSummary.rootMeanSquaredError}")

Coefficients: [-56613.4178409177,-57030.23495583082,13971.999398905327,4527.086795626515,5377.940581159693,-44790.17973643159,40087.54309236055,78382.6392426935,6880.451763891173,16127.459009091863,421.59603616919736,-70125.95532935184,-81484.00811410286,-46236.50973308891,-45419.16119281148], Intercept: -2342961.214869745
RMSE: 68207.48339945609


lr = linReg_e4c7c9eb0938
lrModel = linReg_e4c7c9eb0938
trainingSummary = org.apache.spark.ml.regression.LinearRegressionTrainingSummary@42bf2ac9


org.apache.spark.ml.regression.LinearRegressionTrainingSummary@42bf2ac9

Now, use `RegressionEvaluator` to measure the root-mean-square-erroe (RMSE) of the model on the test dataset.

In [27]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.evaluation.RegressionEvaluator

// make predictions on the test data
val predictions = lrModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator()
                    .setLabelCol("label")
                    .setPredictionCol("prediction")
                    .setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 201417.7826680909| 94600.0|[-62.065401082150...|
|102833.45250053937| 85800.0|[-62.040445150874...|
|184106.64606994344| 76100.0|[-62.015489219599...|
|119394.02364666201| 68400.0|[-61.995524474579...|
| 179417.4987487467|107000.0|[-61.980550915813...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 66606.6870919953


predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_cf1a40edd8b9
rmse = 66606.6870919953


66606.6870919953

## 5.2. Decision tree regression
Repeat what you have done on Regression Model to build a Decision Tree model. Use the `DecisionTreeRegressor` to make a model and then measure its RMSE on the test dataset.

In [28]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.DecisionTreeRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val dt = new DecisionTreeRegressor()
             .setLabelCol("label")
             .setFeaturesCol("features")

// train the model
val dtModel = dt.fit(trainSet)

// make predictions on the test data
val predictions = dtModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator()
                    .setLabelCol("label")
                    .setPredictionCol("prediction")
                    .setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
| 173453.6749090909| 94600.0|[-62.065401082150...|
|144185.68954248365| 85800.0|[-62.040445150874...|
|  135431.654676259| 76100.0|[-62.015489219599...|
| 173453.6749090909| 68400.0|[-61.995524474579...|
| 173453.6749090909|107000.0|[-61.980550915813...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 65965.03824663631


dt = dtr_bd051054383f
dtModel = DecisionTreeRegressionModel (uid=dtr_bd051054383f) of depth 5 with 63 nodes
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_fdad81c537ca
rmse = 65965.03824663631


65965.03824663631

## 5.3. Random forest regression
Let's try the test error on a Random Forest Model. Youcan use the `RandomForestRegressor` to make a Random Forest model.

In [29]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.RandomForestRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val rf = new RandomForestRegressor()
             .setLabelCol("label")
             .setFeaturesCol("features")

// train the model
val rfModel = rf.fit(trainSet)

// make predictions on the test data
val predictions = rfModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator()
                    .setLabelCol("label")
                    .setPredictionCol("prediction")
                    .setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|180450.41943929743| 94600.0|[-62.065401082150...|
|156003.96933102625| 85800.0|[-62.040445150874...|
| 198961.6865302619| 76100.0|[-62.015489219599...|
|162308.94935202197| 68400.0|[-61.995524474579...|
|162491.55406166916|107000.0|[-61.980550915813...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 64812.263206841824


rf = rfr_ac3e6e5698b2
rfModel = RandomForestRegressionModel (uid=rfr_ac3e6e5698b2) with 20 trees
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_6ab0d41eaa75
rmse = 64812.263206841824


64812.263206841824

## 5.4. Gradient-boosted tree regression
Fianlly, we want to build a Gradient-boosted Tree Regression model and test the RMSE of the test data. Use the `GBTRegressor` to build the model.

In [30]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.regression.GBTRegressor
import org.apache.spark.ml.evaluation.RegressionEvaluator

val gb = new GBTRegressor()
             .setLabelCol("label")
             .setFeaturesCol("features")

// train the model
val gbModel = gb.fit(trainSet)

// make predictions on the test data
val predictions = gbModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

// select (prediction, true label) and compute test error
val evaluator = new RegressionEvaluator()
                    .setLabelCol("label")
                    .setPredictionCol("prediction")
                    .setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|110125.25567589799| 94600.0|[-62.065401082150...|
|101912.04648044672| 85800.0|[-62.040445150874...|
| 81965.57146358652| 76100.0|[-62.015489219599...|
|114393.48452975844| 68400.0|[-61.995524474579...|
| 94059.00213108407|107000.0|[-61.980550915813...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 54380.43667449267


gb = gbtr_c574fdbcdaa9
gbModel = GBTRegressionModel (uid=gbtr_c574fdbcdaa9) with 20 trees
predictions = [features: vector, label: double ... 1 more field]
evaluator = regEval_3e1b11a9ca8b
rmse = 54380.43667449267


54380.43667449267

---
# 6. Hyperparameter tuning
An important task in Machie Learning is model selection, or using data to find the best model or parameters for a given task. This is also called tuning. Tuning may be done for individual Estimators such as LinearRegression, or for entire Pipelines which include multiple algorithms, featurization, and other steps. Users can tune an entire Pipeline at once, rather than tuning each element in the Pipeline separately. MLlib supports model selection tools, such as `CrossValidator`. These tools require the following items:
* Estimator: algorithm or Pipeline to tune (`setEstimator`)
* Set of ParamMaps: parameters to choose from, sometimes called a "parameter grid" to search over (`setEstimatorParamMaps`)
* Evaluator: metric to measure how well a fitted Model does on held-out test data (`setEvaluator`)

`CrossValidator` begins by splitting the dataset into a set of folds, which are used as separate training and test datasets. For example with `k=3` folds, `CrossValidator` will generate 3 (training, test) dataset pairs, each of which uses 2/3 of the data for training and 1/3 for testing. To evaluate a particular `ParamMap`, `CrossValidator` computes the average evaluation metric for the 3 Models produced by fitting the Estimator on the 3 different (training, test) dataset pairs. After identifying the best `ParamMap`, `CrossValidator` finally re-fits the Estimator using the best ParamMap and the entire dataset.

Below, use the `CrossValidator` to select the best Random Forest model. To do so, you need to define a grid of parameters. Let's say we want to do the search among the different number of trees (1, 5, and 10), and different tree depth (5, 10, and 15).

---
# 7. Custom transformer
At the end of part two, we added extra columns to the `housing` dataset. Here, we are going to implement a Transformer to do the same task. The Transformer should take the name of two input columns `inputCol1` and `inputCol2`, as well as the name of ouput column `outputCol`. It, then, computes `inputCol1` divided by `inputCol2`, and adds its result as a new column to the dataset. The details of the implemeting a custom Tranfomer is explained [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types). Please read it before before starting to implement it.

First, define the given parameters of the Transformer and implement a method to validate their schemas (`StructType`).

In [32]:
import org.apache.spark.sql.types.{StructField, StructType, DoubleType}
import org.apache.spark.ml.param.{ParamMap, Param, Params}

trait MyParams extends Params {
    final val inputCol1 = new Param[String](this, "inputCol1", "The input column 1")
    final val inputCol2 = new Param[String](this, "inputCol2", "The input column 2")
    final val outputCol = new Param[String](this, "ouputCol", "The output column")
    
  protected def validateAndTransformSchema(schema: StructType): StructType = {

    val ip1 = schema.fieldIndex($(inputCol1))
    val val1 = schema.fields(ip1)

    val ip2 = schema.fieldIndex($(inputCol2))
    val val2 = schema.fields(ip2)  


    if(val1.dataType != DoubleType) {
        throw new Exception(s"Input type ${val1.dataType} did not match input type DoubleType")
    }

    if(val2.dataType != DoubleType) {
        throw new Exception(s"Input type ${val2.dataType} did not match input type DoubleType")
    }
      
    schema.add(StructField($(outputCol), DoubleType, false))

  }
}


defined trait MyParams


Then, extend the class `Transformer`, and implement its setter functions for the input and output columns, and call then `setInputCol1`, `setInputCol2`, and `setOutputCol`. Morever, you need to override the methods `copy`, `transformSchema`, and the `transform`. The details of what you need to cover in these methods is given [here](https://www.oreilly.com/learning/extend-spark-ml-for-your-own-modeltransformer-types).

In [33]:
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.ml.Transformer
import org.apache.spark.ml.param.{ParamMap, Param, Params}
import org.apache.spark.sql.{DataFrame, Dataset}
import org.apache.spark.sql.types.{StructType, StructField, DoubleType}
import org.apache.spark.sql.functions.{col, udf}

class MyTransformer(override val uid: String) extends Transformer with MyParams {
    def this() = this(Identifiable.randomUID("divcolumns"))
    
    def setInputCol1(value: String): MyTransformer = set(inputCol1, value)
    
    def setInputCol2(value: String): MyTransformer = set(inputCol2, value)
    
    def setOutputCol(value: String): MyTransformer = set(outputCol, value)

    override def copy(extra: ParamMap): MyTransformer = {
        defaultCopy(extra)
    }

    override def transformSchema(schema: StructType): StructType = {

        val ip1 = schema.fieldIndex($(inputCol1))
        val val1 = schema.fields(ip1)
      
        val ip2 = schema.fieldIndex($(inputCol2))
        val val2 = schema.fields(ip2)  
      
        if (val1.dataType != DoubleType) {
          throw new Exception(s"Type mismatch for ${val1.dataType} w.r.t expected input type:Double")
        }
    
        if (val2.dataType != DoubleType) {
          throw new Exception(s"Type mismatch for ${val2.dataType} w.r.t expected input type:Double")
        }
      
        schema.add(StructField($(outputCol), DoubleType, false))
        
    }
    
    override def transform(dataset: Dataset[_]): DataFrame = {
        val divider = udf{(column1: Double, column2: Double) => column1/column2}
        dataset.select(col("*"), divider(dataset($(inputCol1)),dataset($(inputCol2))).as($(outputCol)))
  }
}

defined class MyTransformer


Now, an instance of `MyTransformer`, and set the input columns `total_rooms` and `households`, and the output column `rooms_per_household` and run it over the `housing` dataset.

In [34]:
val myTransformer = new MyTransformer().setInputCol1("total_rooms").setInputCol2("households").setOutputCol("rooms_per_household")

val myDataset = myTransformer.transform(housing).select("rooms_per_household").show(5)

+-------------------+
|rooms_per_household|
+-------------------+
|  6.984126984126984|
|  6.238137082601054|
|  8.288135593220339|
| 5.8173515981735155|
|  6.281853281853282|
+-------------------+
only showing top 5 rows



myTransformer = divcolumns_a2dd38aae4d0


myDataset: Unit = ()


divcolumns_a2dd38aae4d0

---
# 8. Custom estimator (predictor)
Now, it's time to implement your own linear regression with gradient descent algorithm as a brand new Estimator. The whole code of the Estimator is given to you, and you do not need to implement anything. It is just a sample that shows how to build a custom Estimator.

The gradient descent update for linear regression is:
$$
w_{i+1} = w_{i} - \alpha_{i} \sum\limits_{j=1}^n (w_i^\top x_j - y_j)x_j
$$

where $i$ is the iteration number of the gradient descent algorithm, and $j$ identifies the observation. Here, $w$ represents an array of weights that is the same size as the array of features and provides a weight for each of the features when finally computing the label prediction in the form:

$$
prediction = w^\top \cdot\ x
$$

where $w$ is the final array of weights computed by the gradient descent, $x$ is the array of features of the observation point and $prediction$ is the label we predict should be associated to this observation.

The given `Helper` class implements the helper methods:
* `dot`: implements the dot product of two vectors and the dot product of a vector and a scalar
* `sum`: implements addition of two vectors
* `fill`: creates a vector of predefined size and initialize it with the predefined value

What you need to do is to implement the methods of the Linear Regresstion class `LR`, which are
* `rmsd`: computes the Root Mean Square Error of a given RDD of tuples of (label, prediction) using the formula:
$$
rmse = \sqrt{\frac{\sum\limits_{i=1}^n (label - prediction)^2}{n}}
$$
* `gradientSummand`: computes the following formula:
$$
gs_{ij} = (w_i^\top x_j - y_j)x_j
$$
* `gradient`: computes the following formula:
$$
gradient = \sum\limits_{j=1}^n gs_{ij}
$$

In [35]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

case class Instance(label: Double, features: Vector)

object Helper extends Serializable {
  def dot(v1: Vector, v2: Vector): Double = {
    val m = Matrices.dense(1, v1.size, v1.toArray)
    m.multiply(v2).values(0)
  }

  def dot(v: Vector, s: Double): Vector = {
    val baseArray = v.toArray.map(vi => vi * s)
    Vectors.dense(baseArray)
  }

  def sumVectors(v1: Vector, v2: Vector): Vector = {
    val baseArray = ((v1.toArray) zip (v2.toArray)).map { case (val1, val2) => val1 + val2 }
    Vectors.dense(baseArray)
  }

  def fillVector(size: Int, fillVal: Double): Vector = Vectors.dense(Array.fill[Double](size)(fillVal));
}

defined class Instance
defined object Helper


In [36]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

class LR() extends Serializable {
  def calcRMSE(labelsAndPreds: RDD[(Double, Double)]): Double = {
    val regressionMetrics = new RegressionMetrics(labelsAndPreds)
    regressionMetrics.rootMeanSquaredError
  }
  
  def gradientSummand(weights: Vector, lp: Instance): Vector = {
    val mult = (Helper.dot(weights, lp.features) - lp.label)
    val seq = (0 to lp.features.size - 1).map(i => lp.features(i) * mult)
    return Vectors.dense(seq.toArray)
  }
  
  def linregGradientDescent(trainData: RDD[Instance], numIters: Int): (Vector, Array[Double]) = {
    val n = trainData.count()
    val d = trainData.take(1)(0).features.size
    var w = Helper.fillVector(d, 0)
    val alpha = 1.0
    val errorTrain = Array.fill[Double](numIters)(0.0)

    for (i <- 0 until numIters) {
      val labelsAndPredsTrain = trainData.map(lp => (lp.label, Helper.dot(w, lp.features)))
      errorTrain(i) = calcRMSE(labelsAndPredsTrain)

      val gradient = trainData.map(lp => gradientSummand(w, lp)).reduce((v1, v2) => Helper.sumVectors(v1, v2))
      val alpha_i = alpha / (n * scala.math.sqrt(i + 1))
      val wAux = Helper.dot(gradient, (-1) * alpha_i)
      w = Helper.sumVectors(w, wAux)
    }
    (w, errorTrain)
  }
}

defined class LR


In [37]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

abstract class MyLinearModel[FeaturesType, Model <: MyLinearModel[FeaturesType, Model]]
  extends PredictionModel[FeaturesType, Model] {
}

class MyLinearModelImpl(override val uid: String, val weights: Vector, val trainingError: Array[Double])
    extends MyLinearModel[Vector, MyLinearModelImpl] {

  override def copy(extra: ParamMap): MyLinearModelImpl = defaultCopy(extra)

  def predict(features: Vector): Double = {
    println("Predicting")
    val prediction = Helper.dot(weights, features)
    prediction
  }
}

defined class MyLinearModel
defined class MyLinearModelImpl


In [38]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._

import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}

abstract class MyLinearRegression[
    FeaturesType,
    Learner <: MyLinearRegression[FeaturesType, Learner, Model],
    Model <: MyLinearModel[FeaturesType, Model]]
  extends Predictor[FeaturesType, Learner, Model] {
}

class MyLinearRegressionImpl(override val uid: String)
    extends MyLinearRegression[Vector, MyLinearRegressionImpl, MyLinearModelImpl] {
  def this() = this(Identifiable.randomUID("linReg"))

  override def copy(extra: ParamMap): MyLinearRegressionImpl = defaultCopy(extra)
  
  def train(dataset: Dataset[_]): MyLinearModelImpl = {
    println("Training")

    val numIters = 10

    val instances: RDD[Instance] = dataset.select(
      col($(labelCol)), col($(featuresCol))).rdd.map {
        case Row(label: Double, features: Vector) =>
          Instance(label, features)
      }

    val (weights, trainingError) = new LR().linregGradientDescent(instances, numIters)

    new MyLinearModelImpl(uid, weights, trainingError)
  }
}

defined class MyLinearRegression
defined class MyLinearRegressionImpl


In [39]:
import org.apache.spark.ml.linalg.Vector
import org.apache.spark.ml.PredictorParams
import org.apache.spark.ml.param.ParamMap
import org.apache.spark.ml.util._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.Dataset
import org.apache.spark.sql.functions._
import org.apache.spark.sql.Row
import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Matrices
import org.apache.spark.mllib.evaluation.RegressionMetrics
import org.apache.spark.ml.{PredictionModel, Predictor}
import org.apache.spark.ml.evaluation.RegressionEvaluator

val lr = new MyLinearRegressionImpl().setLabelCol("label").setFeaturesCol("features")
val model = lr.fit(trainSet)
val predictions = model.transform(trainSet)
predictions.select("prediction", "label", "features").show(5)

// Select (prediction, true label) and compute test error.
val evaluator = new RegressionEvaluator().setLabelCol("label").setPredictionCol("prediction").setMetricName("rmse")
val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

Training
+--------------------+--------+--------------------+
|          prediction|   label|            features|
+--------------------+--------+--------------------+
|-8.70144742470063...|103600.0|[-62.040445150874...|
|-8.68805452053661...| 79000.0|[-62.025471592109...|
|-8.68437365453994...|111400.0|[-62.020480405854...|
|-8.68452109389124...|106700.0|[-62.005506847089...|
|-8.68607339380906...| 50800.0|[-62.005506847089...|
+--------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 8.306968629836763E37


lr = linReg_d8cca776b3be
model = linReg_d8cca776b3be
predictions = [features: vector, label: double ... 1 more field]


evaluator: org.apache.spark.ml.evaluation.Regr...


[features: vector, label: double ... 1 more field]

---
# 9. An End-to-End Classification Test
As the last step, you are given a dataset called `data/ccdefault.csv`. The dataset represents default of credit card clients. It has 30,000 cases and 24 different attributes. More details about the dataset is available at `data/ccdefault.txt`. In this task you should make three models, compare their results and conclude the ideal solution. Here are the suggested steps:
1. Load the data.
2. Carry out some exploratory analyses (e.g., how various features and the target variable are distributed).
3. Train a model to predict the target variable (risk of `default`).
  - Employ three different models (logistic regression, decision tree, and random forest).
  - Compare the models' performances (e.g., AUC).
  - Defend your choice of best model (e.g., what are the strength and weaknesses of each of these models?).
4. What more would you do with this data? Anything to help you devise a better solution?

In [31]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.RegressionEvaluator
import org.apache.spark.ml.tuning.CrossValidator

val paramGrid = new ParamGridBuilder()
                    .addGrid(rf.numTrees, Array(1, 5, 10))
                    .addGrid(rf.maxDepth, Array(5, 10, 15))
                    .build()

val evaluator = new RegressionEvaluator()
                    .setLabelCol("label")
                    .setPredictionCol("prediction")
                    .setMetricName("rmse")

val cv = new CrossValidator()
             .setEstimator(rf)
             .setEvaluator(new RegressionEvaluator)
             .setEstimatorParamMaps(paramGrid)
             .setNumFolds(3)  // Use 3+ in practice
             .setParallelism(2)  // Evaluate up to 2 parameter settings in parallel

val cvModel = cv.fit(trainSet)

val predictions = cvModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

val rmse = evaluator.evaluate(predictions)
println(s"Root Mean Squared Error (RMSE) on test data = $rmse")

+------------------+--------+--------------------+
|        prediction|   label|            features|
+------------------+--------+--------------------+
|          125010.0| 94600.0|[-62.065401082150...|
| 70272.83783783784| 85800.0|[-62.040445150874...|
|          122322.0| 76100.0|[-62.015489219599...|
| 88287.77777777778| 68400.0|[-61.995524474579...|
|111549.16666666666|107000.0|[-61.980550915813...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 52565.887163502244


paramGrid = 


Array({
	rfr_ac3e6e5698b2-maxDepth: 5,
	rfr_ac3e6e5698b2-numTrees: 1
}, {
	rfr_ac3e6e5698b2-maxDepth: 5,
	rfr_ac3e6e5698b2-numTrees: 5
}, {
	rfr_ac3e6e5698b2-maxDepth: 5,
	rfr_ac3e6e5698b2-numTrees: 10
}, {
	rfr_ac3e6e5698b2-maxDepth: 10,
	rfr_ac3e6e5698b2-numTrees: 1
}, {
	rfr_ac3e6e5698b2-maxDepth: 10,
	rfr_ac3e6e5698b2-numTrees: 5
}, {
	rfr_ac3e6e5698b2-maxDepth: 10,
	rfr_ac3e6e5698b2-numTrees: 10
}, {
	rfr_ac3e6e5698b2-maxDepth: 15,
	rfr_ac3e6e5698b2-numTrees: 1
}, {
	rfr_ac3e6e5698b2-maxDepth: 15,
	rfr_ac3e6e5698b2-numTrees: 5
}, {
	rfr_ac3e6e5698b2-maxDepth: 15,
	rfr_ac3e...


## Dataset

This research aimed at the case of customers' default payments in Taiwan and compares the predictive accuracy of probability of default among six data mining methods. From the perspective of risk management, the result of predictive accuracy of the estimated probability of default will be more valuable than the binary result of classification - credible or not credible clients. Because the real probability of default is unknown, this study presented the novel "Sorting Smoothing Method" to estimate the real probability of default. With the real probability of default as the response variable (Y), and the predictive probability of default as the independent variable (X), the simple linear regression result (Y = A + BX) shows that the forecasting model produced by artificial neural network has the highest coefficient of determination; its regression intercept (A) is close to zero, and regression coefficient (B) to one. Therefore, among the six data mining techniques, artificial neural network is the only one that can accurately estimate the real probability of default.

This research employed a binary variable, default payment (Yes = 1, No = 0), as the response variable. This study reviewed the literature and used the following 23 variables as explanatory variables: 
1. X1: Amount of the given credit (NT dollar): it includes both the individual consumer credit and his/her family (supplementary) credit. 
2. X2: Gender (1 = male; 2 = female). 
3. X3: Education (1 = graduate school; 2 = university; 3 = high school; 4 = others). 
4. X4: Marital status (1 = married; 2 = single; 3 = others). 
5. X5: Age (year). 
6. X6 - X11: History of past payment. We tracked the past monthly payment records (from April to September, 2005) as follows: X6 = the repayment status in September, 2005; X7 = the repayment status in August, 2005; . . .;X11 = the repayment status in April, 2005. The measurement scale for the repayment status is: -1 = pay duly; 1 = payment delay for one month; 2 = payment delay for two months; . . .; 8 = payment delay for eight months; 9 = payment delay for nine months and above. 
7. X12-X17: Amount of bill statement (NT dollar). X12 = amount of bill statement in September, 2005; X13 = amount of bill statement in August, 2005; . . .; X17 = amount of bill statement in April, 2005. 
8. X18-X23: Amount of previous payment (NT dollar). X18 = amount paid in September, 2005; X19 = amount paid in August, 2005; . . .;
9. X23 = amount paid in April, 2005. 

The label is the Default: 1 or 0


In [1]:
val creditcard = spark.read.format("csv")
                      .option("header", "true")
                      .option("inferSchema", "true")
                      .load("data/ccdefault.csv")

Intitializing Scala interpreter ...

Spark Web UI available at http://130.229.189.178:4040
SparkContext available as 'sc' (version = 2.4.4, master = local[*], app id = local-1574507470811)
SparkSession available as 'spark'


creditcard: org.apache.spark.sql.DataFrame = [ID: int, LIMIT_BAL: int ... 23 more fields]


In [2]:
creditcard.printSchema()

root
 |-- ID: integer (nullable = true)
 |-- LIMIT_BAL: integer (nullable = true)
 |-- SEX: integer (nullable = true)
 |-- EDUCATION: integer (nullable = true)
 |-- MARRIAGE: integer (nullable = true)
 |-- AGE: integer (nullable = true)
 |-- PAY_0: integer (nullable = true)
 |-- PAY_2: integer (nullable = true)
 |-- PAY_3: integer (nullable = true)
 |-- PAY_4: integer (nullable = true)
 |-- PAY_5: integer (nullable = true)
 |-- PAY_6: integer (nullable = true)
 |-- BILL_AMT1: integer (nullable = true)
 |-- BILL_AMT2: integer (nullable = true)
 |-- BILL_AMT3: integer (nullable = true)
 |-- BILL_AMT4: integer (nullable = true)
 |-- BILL_AMT5: integer (nullable = true)
 |-- BILL_AMT6: integer (nullable = true)
 |-- PAY_AMT1: integer (nullable = true)
 |-- PAY_AMT2: integer (nullable = true)
 |-- PAY_AMT3: integer (nullable = true)
 |-- PAY_AMT4: integer (nullable = true)
 |-- PAY_AMT5: integer (nullable = true)
 |-- PAY_AMT6: integer (nullable = true)
 |-- DEFAULT: integer (nullable = tru

Let's print the first 12 columns of the dataset to see how it looks like:

In [12]:
import org.apache.spark.sql.functions._

var some_columns: Array[String] = creditcard.columns.take(12)
var some_columns_5: Array[String] = creditcard.columns.take(5)
creditcard.select(some_columns.head, some_columns.tail: _*).show(10)

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|
|  2|   120000|  2|        2|       2| 26|   -1|    2|    0|    0|    0|    2|
|  3|    90000|  2|        2|       2| 34|    0|    0|    0|    0|    0|    0|
|  4|    50000|  2|        2|       1| 37|    0|    0|    0|    0|    0|    0|
|  5|    50000|  1|        2|       1| 57|   -1|    0|   -1|    0|    0|    0|
|  6|    50000|  1|        1|       2| 37|    0|    0|    0|    0|    0|    0|
|  7|   500000|  1|        1|       2| 29|    0|    0|    0|    0|    0|    0|
|  8|   100000|  2|        2|       2| 23|    0|   -1|   -1|    0|    0|   -1|
|  9|   140000|  2|        3|       1| 28|    0|    0|    2|    0|    0|    0|
| 10|    20000|  1|        3|       2| 35|   -2|   -

import org.apache.spark.sql.functions._
some_columns: Array[String] = Array(ID, LIMIT_BAL, SEX, EDUCATION, MARRIAGE, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6)
some_columns_5: Array[String] = Array(ID, LIMIT_BAL, SEX, EDUCATION, MARRIAGE)


### Check for unique distinct ids

In [4]:
print("Rows are " + creditcard.count() + " They should be as many as the ids, that are:\n")
creditcard.select(max("ID")).show()

Rows are 30000 They should be as many as the ids, that are:
+-------+
|max(ID)|
+-------+
|  30000|
+-------+



## Dataset balancing check

We notice there are many categorical features (already numerical). Therefore we will explore if some of these categories are represented equally.
We check for male and females in the dataset, as well as the males with default and females with defaults.

In [5]:
creditcard.createOrReplaceTempView("df")
spark.sql("SELECT sex, count(*), count(*)*100/(SELECT count(*) FROM df as df2) as percentage FROM df GROUP BY sex").show()

+---+--------+------------------+
|sex|count(1)|        percentage|
+---+--------+------------------+
|  1|   11888|39.626666666666665|
|  2|   18112|60.373333333333335|
+---+--------+------------------+



Dataset seems to be a little bit unbalanced:
If we select ones being females and males and compare their default rate, we also notice a little unbalancing in the 2 classes M/F:

In [6]:
creditcard.createOrReplaceTempView("df")
spark.sql("SELECT sex, count(*)*100/(SELECT count(*) from df as df2 where df2.sex == sex) as percentage_default FROM df WHERE default == 1 GROUP BY sex").show()

+---+------------------+
|sex|percentage_default|
+---+------------------+
|  1| 9.576666666666666|
|  2|12.543333333333333|
+---+------------------+



## Correlation of the features

As done before, let's vectorize the columns of the dataframe and get their correlation scores

In [7]:
import org.apache.spark.ml.feature.VectorAssembler

// get input multiple columns, set as output the name of the column output that will represent all the features of each row 
// (spark row formatting [a,b,c,d,...])
val va = new VectorAssembler()
             .setInputCols(creditcard.columns)
             .setOutputCol("new_vect")

val CreditCardAttrs = va.transform(creditcard)

CreditCardAttrs.show(5)


+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+--------------------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|DEFAULT|            new_vect|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+--------------------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|      1|[1.0,20000.0,2.0,...|
|  2|   120000|  2|        2|       2| 26|   -1|    2|    0|    0|    0|    2|     2682|     1725|     2682|    

import org.apache.spark.ml.feature.VectorAssembler
va: org.apache.spark.ml.feature.VectorAssembler = vecAssembler_a8e1f83c3476
CreditCardAttrs: org.apache.spark.sql.DataFrame = [ID: int, LIMIT_BAL: int ... 24 more fields]


In [8]:
import org.apache.spark.ml.linalg.Matrix
import org.apache.spark.ml.stat.Correlation
import org.apache.spark.sql.Row

val Row(coeff: Matrix) = Correlation.corr(CreditCardAttrs, "new_vect").head

println(s"The standard correlation coefficient:\n ${coeff}")

The standard correlation coefficient:
 1.0                     0.026179157746199246  ... (25 total)
0.026179157746199246    1.0                   ...
0.018497494176216264    0.024755235111645853  ...
0.03917673481156041     -0.2191606982292233   ...
-0.02907942531876594    -0.10813941027800818  ...
0.018677734490348097    0.14471279755736938   ...
-0.030574932272971585   -0.27121433213471846  ...
-0.01121486826234242    -0.2963821009964793   ...
-0.01849355375769931    -0.2861229539031139   ...
-0.0027348897819585508  -0.2674600096393908   ...
-0.02219923187001884    -0.24941139482410882  ...
-0.02027013406721332    -0.23519539948542625  ...
0.019388673917612936    0.28542986496499084   ...
0.017981897457702494    0.27831436399776227   ...
0.02435405001883995     0.283235783581682     ...
0.04035060294250819     0.2939876237159879    ...
0.016705303156357654    0.29556233765823553   ...
0.016729618316252175    0.2903889506479471    ...
0.009742443464324053    0.19523591523220815   ...


import org.apache.spark.ml.linalg.Matrix
import org.apache.spark.ml.stat.Correlation
import org.apache.spark.sql.Row
coeff: org.apache.spark.ml.linalg.Matrix =
1.0                     0.026179157746199246  ... (25 total)
0.026179157746199246    1.0                   ...
0.018497494176216264    0.024755235111645853  ...
0.03917673481156041     -0.2191606982292233   ...
-0.02907942531876594    -0.10813941027800818  ...
0.018677734490348097    0.14471279755736938   ...
-0.030574932272971585   -0.27121433213471846  ...
-0.01121486826234242    -0.2963821009964793   ...
-0.01849355375769931    -0.2861229539031139   ...
-0.0027348897819585508  -0.2674600096393908   ...
-0.02219923187001884    -0.24941139482410882  ...
-0.02027013406721332    -0.23519539948542625  ...
0.01938867...

## Preprocessing: fix the categorical features

We observe that SEX, EDUCATION, and MARRIAGE are numerical but must be considered as categorical features

Let's print the distinct elements of each

In [9]:
// count disting element of categorical columns
creditcard.createOrReplaceTempView("df")
spark.sql("SELECT count(DISTINCT SEX), count(DISTINCT EDUCATION), count(DISTINCT MARRIAGE) FROM df").show()

+-------------------+-------------------------+------------------------+
|count(DISTINCT SEX)|count(DISTINCT EDUCATION)|count(DISTINCT MARRIAGE)|
+-------------------+-------------------------+------------------------+
|                  2|                        7|                       4|
+-------------------+-------------------------+------------------------+



## One hot encoder
* Since SEX has only 2 values, we will treat is as binary (0 or 1), we will scale by 1
* for the others 2, (Education and Marriage), we need to one-hot-encode their values

In [10]:
import org.apache.spark.ml.feature.OneHotEncoderEstimator

val encoder = new OneHotEncoderEstimator()
                  .setInputCols(Array("EDUCATION", "MARRIAGE"))
                  .setOutputCols(Array("EDUCATION_vec", "MARRIAGE_vec"))
val ohCC = encoder.fit(creditcard).transform(creditcard)

ohCC.show(5)

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+-------------+-------------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|DEFAULT|EDUCATION_vec| MARRIAGE_vec|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+-------------+-------------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0|      1|(6,[2],[1.0])|(3,[1],[1.0])|
|  2|   120000|  2|        2|       2| 26|   -1|    2|    0|    0|    0|    2|     2

import org.apache.spark.ml.feature.OneHotEncoderEstimator
encoder: org.apache.spark.ml.feature.OneHotEncoderEstimator = oneHotEncoder_4eee116840d9
ohCC: org.apache.spark.sql.DataFrame = [ID: int, LIMIT_BAL: int ... 25 more fields]


In [11]:
// rename DEFAULT as label
val renamedCC = ohCC.withColumnRenamed("DEFAULT","label")

// label columns
val colLabel = "label"

// SEX, EDUCATION and MARRIAGE are categorical
// Let's leave only numerical columns
val colNum_1 = renamedCC.columns.filter(_ != colLabel).filter(_ != "SEX").filter(_ != "EDUCATION").filter(_ != "MARRIAGE").filter(_ != "MARRIAGE_vec").filter(_ != "EDUCATION_vec")
//val colNum = colNum_1.columns.filter(col => !(col.isin(colCat:_*)))

//TODO Automatize the filtering in the list: I couldn't make it

renamedCC: org.apache.spark.sql.DataFrame = [ID: int, LIMIT_BAL: int ... 25 more fields]
colLabel: String = label
colNum_1: Array[String] = Array(ID, LIMIT_BAL, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6)


## Scaling

In [14]:
import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}

val va = new VectorAssembler()
             .setInputCols(colNum_1)
             .setOutputCol("feat_intermediate")

val featuredCC = va.transform(renamedCC)

val scaler = new StandardScaler()
                 .setInputCol("feat_intermediate")
                 .setOutputCol("scaledFeatures")
val scaledCC = scaler.fit(featuredCC).transform(featuredCC)

scaledCC.show(5)

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+-------------+-------------+--------------------+--------------------+
| ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|label|EDUCATION_vec| MARRIAGE_vec|   feat_intermediate|      scaledFeatures|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+-------------+-------------+--------------------+--------------------+
|  1|    20000|  2|        2|       1| 24|    2|    2|   -1|   -1|   -2|   -2|     3913|     3102|      689|        0|        0|        0|       0|     689|       0|       0|       0|       0| 

import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}
va: org.apache.spark.ml.feature.VectorAssembler = vecAssembler_74a35e1d013b
featuredCC: org.apache.spark.sql.DataFrame = [ID: int, LIMIT_BAL: int ... 26 more fields]
scaler: org.apache.spark.ml.feature.StandardScaler = stdScal_4e091a5322ed
scaledCC: org.apache.spark.sql.DataFrame = [ID: int, LIMIT_BAL: int ... 27 more fields]


## Missing value check

In [15]:
for (c <- colNum_1){
    var count = renamedCC.filter( renamedCC(c).isNull || renamedCC(c)==="").count()
    print("Column " + c + " has " + count + " missing values!\n")
}

Column ID has 0 missing values!
Column LIMIT_BAL has 0 missing values!
Column AGE has 0 missing values!
Column PAY_0 has 0 missing values!
Column PAY_2 has 0 missing values!
Column PAY_3 has 0 missing values!
Column PAY_4 has 0 missing values!
Column PAY_5 has 0 missing values!
Column PAY_6 has 0 missing values!
Column BILL_AMT1 has 0 missing values!
Column BILL_AMT2 has 0 missing values!
Column BILL_AMT3 has 0 missing values!
Column BILL_AMT4 has 0 missing values!
Column BILL_AMT5 has 0 missing values!
Column BILL_AMT6 has 0 missing values!
Column PAY_AMT1 has 0 missing values!
Column PAY_AMT2 has 0 missing values!
Column PAY_AMT3 has 0 missing values!
Column PAY_AMT4 has 0 missing values!
Column PAY_AMT5 has 0 missing values!
Column PAY_AMT6 has 0 missing values!


## Finally make up the dataset

In [16]:
// numerical columns
val finalCC = scaledCC.drop("EDUCATION", "MARRIAGE")
finalCC.printSchema()

// label columns
val colLabel = "label"

// numerical columns
val newColNum = finalCC.columns.filter(_ != colLabel).filter(_ != "SEX").filter(_ != "EDUCATION").filter(_ != "MARRIAGE")

print(newColNum)
val va2 = new VectorAssembler()
              .setInputCols(newColNum)
              .setOutputCol("features")

val dataset = va2.transform(finalCC).select("features", "label")


root
 |-- ID: integer (nullable = true)
 |-- LIMIT_BAL: integer (nullable = true)
 |-- SEX: integer (nullable = true)
 |-- AGE: integer (nullable = true)
 |-- PAY_0: integer (nullable = true)
 |-- PAY_2: integer (nullable = true)
 |-- PAY_3: integer (nullable = true)
 |-- PAY_4: integer (nullable = true)
 |-- PAY_5: integer (nullable = true)
 |-- PAY_6: integer (nullable = true)
 |-- BILL_AMT1: integer (nullable = true)
 |-- BILL_AMT2: integer (nullable = true)
 |-- BILL_AMT3: integer (nullable = true)
 |-- BILL_AMT4: integer (nullable = true)
 |-- BILL_AMT5: integer (nullable = true)
 |-- BILL_AMT6: integer (nullable = true)
 |-- PAY_AMT1: integer (nullable = true)
 |-- PAY_AMT2: integer (nullable = true)
 |-- PAY_AMT3: integer (nullable = true)
 |-- PAY_AMT4: integer (nullable = true)
 |-- PAY_AMT5: integer (nullable = true)
 |-- PAY_AMT6: integer (nullable = true)
 |-- label: integer (nullable = true)
 |-- EDUCATION_vec: vector (nullable = true)
 |-- MARRIAGE_vec: vector (nullable =

finalCC: org.apache.spark.sql.DataFrame = [ID: int, LIMIT_BAL: int ... 25 more fields]
colLabel: String = label
newColNum: Array[String] = Array(ID, LIMIT_BAL, AGE, PAY_0, PAY_2, PAY_3, PAY_4, PAY_5, PAY_6, BILL_AMT1, BILL_AMT2, BILL_AMT3, BILL_AMT4, BILL_AMT5, BILL_AMT6, PAY_AMT1, PAY_AMT2, PAY_AMT3, PAY_AMT4, PAY_AMT5, PAY_AMT6, EDUCATION_vec, MARRIAGE_vec, feat_intermediate, scaledFeatures)
va2: org.apache.spark.ml.feature.VectorAssembler = vecAssembler_d7d50ef28f50
dataset: org.apache.spark.sql.DataFrame = [features: vector, label: int]


## Dataset Creation

In [36]:
val Array(trainSet, testSet) = dataset.randomSplit(Array(0.8,0.2))

trainSet: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [features: vector, label: int]
testSet: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [features: vector, label: int]


## Logistic Regression

#### TODO how it works

In [19]:
import org.apache.spark.ml.classification.LogisticRegression

val lr = new LogisticRegression()
             .setLabelCol("label")
             .setFeaturesCol("features")
             .setMaxIter(30)
             .setRegParam(0.01)
             .setElasticNetParam(0)

// Fit the model
val lrModel = lr.fit(trainSet)

val predictions = lrModel.transform(testSet)

val trainingSummary = lrModel.summary

predictions.select("prediction", "rawPrediction", "label", "features")

// Print the coefficients and intercept for logistic regression
println(s"Coefficients: ${lrModel.coefficients} Intercept: ${lrModel.intercept}")

// Cast the summary to a BinaryLogisticRegressionSummary 
val binarySummary = lrModel.binarySummary

val roc = binarySummary.roc
roc.show()
println(s"areaUnderROC: ${binarySummary.areaUnderROC}")

Coefficients: [-2.603833508239152E-7,-2.223136465908265E-7,0.0021738413680298546,0.18864931804103297,0.019121081387174273,0.027800694457504303,0.006865212124264992,0.012757767616027336,0.0076288961072742585,-9.734168387497988E-7,-1.5141282517334898E-9,3.673541852966228E-7,3.365820373722798E-8,1.1364272222995615E-7,8.85720470130487E-8,-3.5118796081713335E-6,-3.156212268413733E-6,-1.1914992780785994E-6,-1.1903384069957932E-6,-1.1597434852979677E-6,-7.50221733048132E-7,-2.1715234282225038,0.09356700633318413,-0.009375359381013035,-0.022013532843192497,-0.9236259706985364,-1.0013869492705907,-0.9393260364532497,0.09488630807286658,-0.09955796474616674,-2.603833508239152E-7,-2.223136465908265E-7,0.0021738413680298546,0.18864931804103297,0.019121081387174273,0.027800694457504303,0.006865212124264992,0.012757767616027336,0.0076288961072742585,-9.734168387497988E-7,-1.5141282517334898E-9,3.673541852966228E-7,3.365820373722798E-8,1.1364272222995615E-7,8.85720470130487E-8,-3.5118796081713335E-6,

import org.apache.spark.ml.classification.LogisticRegression
lr: org.apache.spark.ml.classification.LogisticRegression = logreg_26861e49621c
lrModel: org.apache.spark.ml.classification.LogisticRegressionModel = LogisticRegressionModel: uid = logreg_26861e49621c, numClasses = 2, numFeatures = 72
predictions: org.apache.spark.sql.DataFrame = [features: vector, label: int ... 3 more fields]
trainingSummary: org.apache.spark.ml.classification.LogisticRegressionTrainingSummary = org.apache.spark.ml.classification.BinaryLogisticRegressionTrainingSummaryImpl@22735b65
binarySummary: org.apache.spark.ml.classification.BinaryLogisticRegressionTrainingSummary = org.apache.spark.ml.classification.BinaryLogisticRegressionTrainingSummaryImpl@22735b65
roc: org.apache.spark.sql.DataFrame = [FPR: ...

## Decision Tree

#### TODO how it works

In [31]:
import org.apache.spark.ml.Pipeline
import org.apache.spark.ml.classification.DecisionTreeClassificationModel
import org.apache.spark.ml.classification.DecisionTreeClassifier
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
import org.apache.spark.ml.feature.{IndexToString, StringIndexer, VectorIndexer}

val labelIndexer = new StringIndexer()
                       .setInputCol("label")
                       .setOutputCol("indexedLabel")
                       .fit(dataset)

val featureIndexer = new VectorIndexer()
                         .setInputCol("features")
                         .setOutputCol("indexedFeatures")
                         .setMaxCategories(2)
                         .fit(dataset)

val dt = new DecisionTreeClassifier()
             .setLabelCol("indexedLabel")
             .setFeaturesCol("indexedFeatures")

val labelConverter = new IndexToString()
                         .setInputCol("prediction")
                         .setOutputCol("predictedLabel")
                         .setLabels(labelIndexer.labels)

val pipeline = new Pipeline()
                   .setStages(Array(labelIndexer, featureIndexer, dt, labelConverter))

val dtModel = pipeline.fit(trainSet)

val predictions = dtModel.transform(testSet)
predictions.select("prediction", "label", "features")

val evaluator = new BinaryClassificationEvaluator()
                    .setLabelCol("indexedLabel")
                    .setRawPredictionCol("prediction")
                    .setMetricName("areaUnderROC")

val areaUnderROC = evaluator.evaluate(predictions)
println(s"areaUnderROC: ${areaUnderROC}")

areaUnderROC: 0.6609917428300444


import org.apache.spark.ml.Pipeline
import org.apache.spark.ml.classification.DecisionTreeClassificationModel
import org.apache.spark.ml.classification.DecisionTreeClassifier
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
import org.apache.spark.ml.feature.{IndexToString, StringIndexer, VectorIndexer}
labelIndexer: org.apache.spark.ml.feature.StringIndexerModel = strIdx_877201dcb2d4
featureIndexer: org.apache.spark.ml.feature.VectorIndexerModel = vecIdx_43cb7d6e2a77
dt: org.apache.spark.ml.classification.DecisionTreeClassifier = dtc_346ce07cc8f4
labelConverter: org.apache.spark.ml.feature.IndexToString = idxToStr_6d36112e2b89
pipeline: org.apache.spark.ml.Pipeline = pipeline_f17b10554bf2
dtModel: org.apache.spark.ml.PipelineModel = pipeline_f17b10554bf2
p...

## Random Forest

#### TODO how it works

In [32]:
import org.apache.spark.ml.Pipeline
import org.apache.spark.ml.classification.{RandomForestClassificationModel, RandomForestClassifier}
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
import org.apache.spark.ml.feature.{IndexToString, StringIndexer, VectorIndexer}

val labelIndexer = new StringIndexer()
                       .setInputCol("label")
                       .setOutputCol("indexedLabel")
                       .fit(dataset)

val featureIndexer = new VectorIndexer()
                         .setInputCol("features")
                         .setOutputCol("indexedFeatures")
                         .setMaxCategories(2)
                         .fit(dataset)

val rf = new RandomForestClassifier()
             .setLabelCol("indexedLabel")
             .setFeaturesCol("indexedFeatures")
             .setNumTrees(10)
             .setSubsamplingRate(0.8)
             .setFeatureSubsetStrategy("auto")

val labelConverter = new IndexToString()
                         .setInputCol("prediction")
                         .setOutputCol("predictedLabel")
                         .setLabels(labelIndexer.labels)

val pipeline = new Pipeline()
                   .setStages(Array(labelIndexer, featureIndexer, rf, labelConverter))

val rfModel = pipeline.fit(trainSet)

val predictions = rfModel.transform(testSet)

predictions.select("predictedLabel", "label", "features")

// Evaluation of the AUC
val evaluator = new BinaryClassificationEvaluator()
                    .setLabelCol("indexedLabel")
                    .setRawPredictionCol("prediction")
                    .setMetricName("areaUnderROC")

val areaUnderROC = evaluator.evaluate(predictions)
println(s"areaUnderROC: ${areaUnderROC}")

areaUnderROC: 0.6552759530301887


import org.apache.spark.ml.Pipeline
import org.apache.spark.ml.classification.{RandomForestClassificationModel, RandomForestClassifier}
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
import org.apache.spark.ml.feature.{IndexToString, StringIndexer, VectorIndexer}
labelIndexer: org.apache.spark.ml.feature.StringIndexerModel = strIdx_2f9421e01b78
featureIndexer: org.apache.spark.ml.feature.VectorIndexerModel = vecIdx_e5b1b600584e
rf: org.apache.spark.ml.classification.RandomForestClassifier = rfc_82be3b702be1
labelConverter: org.apache.spark.ml.feature.IndexToString = idxToStr_da89ea000a6e
pipeline: org.apache.spark.ml.Pipeline = pipeline_c1c4bbc569d3
rfModel: org.apache.spark.ml.PipelineModel = pipeline_c1c4bbc569d3
predictions: org.apache.spark.sql.DataFra...

Up to now, the Logistic Regression Classifier is the model that performs the best. 
- The decision tree is performing with a lower perfomance, this probaly is due to overfitting as Decision Trees are known to lack of generalization.
- With random Forest we will be able to solve the overfitting problem by adding multiple subsaples and trees. 
- However, by exploring the hyperparameter combinations of the Random Forest, it is possible to notice that it is very sensible to changes. 

For this reason, here we perform a CrossValidation and some combination of these hyperparameters.

In [ ]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
import org.apache.spark.ml.tuning.CrossValidator

val rf = new RandomForestClassifier()
             .setLabelCol("label")
             .setFeaturesCol("features")
             .setFeatureSubsetStrategy("auto")

val paramGrid = new ParamGridBuilder()
                    .addGrid(rf.subsamplingRate, Array(0.5, 0.6, 0.7, 0.8, 0.9))
                    .addGrid(rf.numTrees, Array(10,20,30,40))
                    .addGrid(rf.maxDepth, Array(5, 10, 15))
                    .build()

val evaluator = new BinaryClassificationEvaluator()
                    .setLabelCol("label")
                    .setRawPredictionCol("features")
                    .setMetricName("areaUnderROC")

val cv = new CrossValidator()
             .setEstimator(rf)
             .setEvaluator(new BinaryClassificationEvaluator)
             .setEstimatorParamMaps(paramGrid)
             .setNumFolds(3)  // Use 3+ in practice
             .setParallelism(2)  // Evaluate up to 2 parameter settings in parallel

val cvModel = cv.fit(trainSet)

val predictions = cvModel.transform(testSet)
predictions.select("predictedLabel", "label", "features").show(5)

val areaUnderROC = evaluator.evaluate(predictions)
println(s"areaUnderROC on test data = $areaUnderROC")

 ## Comment on results TODO

## Gradient-Boosted Tree Classifier

#### TODO how it works

In [87]:
import org.apache.spark.ml.Pipeline
import org.apache.spark.ml.classification.{GBTClassificationModel, GBTClassifier}
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator
import org.apache.spark.ml.feature.{IndexToString, StringIndexer, VectorIndexer}

val labelIndexer = new StringIndexer()
                       .setInputCol("label")
                       .setOutputCol("indexedLabel")
                       .fit(dataset)

val featureIndexer = new VectorIndexer()
                         .setInputCol("features")
                         .setOutputCol("indexedFeatures")
                         .setMaxCategories(7)
                         .fit(dataset)

val gbt = new GBTClassifier()
              .setLabelCol("indexedLabel")
              .setFeaturesCol("indexedFeatures")
              .setMaxIter(10)

val labelConverter = new IndexToString()
                         .setInputCol("prediction")
                         .setOutputCol("predictedLabel")
                         .setLabels(labelIndexer.labels)

val pipeline = new Pipeline()
                   .setStages(Array(labelIndexer, featureIndexer, rf, labelConverter))

val rfModel = pipeline.fit(trainSet)

val predictions = rfModel.transform(testSet)

predictions.select("predictedLabel", "label", "features")

val evaluator = new BinaryClassificationEvaluator()
                    .setLabelCol("indexedLabel")
                    .setRawPredictionCol("prediction")
                    .setMetricName("areaUnderROC")

val areaUnderROC = evaluator.evaluate(predictions)
println(s"areaUnderROC: ${areaUnderROC}")

+--------------+-----+--------------------+
|predictedLabel|label|            features|
+--------------+-----+--------------------+
|             1|    1|(72,[0,1,2,3,4,5,...|
|             1|    0|(72,[0,1,2,3,4,5,...|
|             0|    0|(72,[0,1,2,3,4,5,...|
|             0|    0|(72,[0,1,2,3,4,5,...|
|             0|    0|(72,[0,1,2,3,4,5,...|
+--------------+-----+--------------------+
only showing top 5 rows

areaUnderROC: 0.6475916228766022


labelIndexer = strIdx_729be8079084
featureIndexer = vecIdx_181af8f71704
gbt = gbtc_a1cd0e8fa153
labelConverter = idxToStr_6b3d6caf38b1
pipeline = pipeline_ebd5813135ef
rfModel = pipeline_ebd5813135ef
predictions = [features: vector, label: ...


[features: vector, label: ...

For the same reasons of RandomForest, lt's perform a CrossValidation on the GBT hyperparamers

In [ ]:
// TODO: Replace <FILL IN> with appropriate code

import org.apache.spark.ml.tuning.ParamGridBuilder
import org.apache.spark.ml.evaluation.BinaryClassificationEvaluator
import org.apache.spark.ml.tuning.CrossValidator

val gbt = new GBTClassifier()
              .setLabelCol("label")
              .setFeaturesCol("features")

val paramGrid = new ParamGridBuilder()
                    .addGrid(gbt.subsamplingRate, Array(0.5, 0.6, 0.7, 0.8, 0.9))
                    .addGrid(gbt.numTrees, Array(10,20,30,40))
                    .addGrid(gbt.maxDepth, Array(5, 10, 15))
                    .build()

val evaluator = new BinaryClassificationEvaluator()
                    .setLabelCol("label")
                    .setRawPredictionCol("features")
                    .setMetricName("areaUnderROC")

val cv = new CrossValidator()
             .setEstimator(gbt)
             .setEvaluator(new BinaryClassificationEvaluator)
             .setEstimatorParamMaps(paramGrid)
             .setNumFolds(3)  // Use 3+ in practice
             .setParallelism(2)  // Evaluate up to 2 parameter settings in parallel

val cvModel = cv.fit(trainSet)

val predictions = cvModel.transform(testSet)
predictions.select("predictedLabel", "label", "features").show(5)

val areaUnderROC = evaluator.evaluate(predictions)
println(s"areaUnderROC on test data = $areaUnderROC")

## Comment on results